Extended LayoutStatistics

Sharif000123 opened the pull request:
Various functions where added from several papers, to extend LayoutStatistics.cpp. Links of the papers can be found in Layoutstatistics.h, in the function descriptions.

Sharif000123 committed:
Extended LayoutStatistics

Sharif000123 committed:
Extended LayoutStatistics

Sharif000123 committed:
Merge branch 'master' of https://github.com/ogdf/ogdf into layoutstatistics

Sharif000123 committed:
Corrected Bracet

Sharif000123 committed:
Corrected Bracet

Sharif000123 committed:
Removed whitspaces in end of lines in Layoutstatistics.h for the style check

Sharif000123 committed:
Removed whitspaces in end of lines in Layoutstatistics.h for the style check

Sharif000123 committed:
Fixed spacing, for doxygen style check

Sharif000123 committed:
Fixed spacing, for doxygen style check

Sharif000123 committed:
Removed "LayoutStatistics::" in function declarations, as merge check complained

Sharif000123 committed:
Removed "LayoutStatistics::" in function declarations, as merge check complained

Sharif000123 committed:
Had to clang format again

Sharif000123 committed:
Had to clang format again

N-Coder reviewed (changes requested):
Hey, thanks for submitting this! Please see my below comments (and fix the related error in CI) and also provide some unit tests that cover as much as possible of the new code, then I'd be happy to merge.

Cheers,
Niko

Sharif000123 committed:
Removed maingraph input everywhere, changed gabrielRatio to recieve an empty graph reference and will assign a gabrielgraph to that graph, moved some functions in the code in a meaningful way (as asked), removed the #Layoutstatistics include, used GraphAttributes::BoundingBox

Sharif000123 committed:
Removed maingraph input everywhere, changed gabrielRatio to recieve an empty graph reference and will assign a gabrielgraph to that graph, moved some functions in the code in a meaningful way (as asked), removed the #Layoutstatistics include, used GraphAttributes::BoundingBox

Sharif000123 committed:
.

Sharif000123 committed:
.

Sharif000123 committed:
Fixed casts, a return placement, and a function declaration

Sharif000123 committed:
Fixed casts, a return placement, and a function declaration

Sharif000123 committed:
Removed some newlines, I guess this should fix the code style check

Sharif000123 committed:
Removed some newlines, I guess this should fix the code style check

Sharif000123 committed:
Added one new line in end of code

Sharif000123 committed:
Added one new line in end of code

Sharif000123 committed:
Uncommented and removed some variables with test complaints, that weren't used enough. Uncommented variables may be commented-in again if needed

Sharif000123 committed:
Uncommented and removed some variables with test complaints, that weren't used enough. Uncommented variables may be commented-in again if needed

Sharif000123 committed:
Formatted, one line was not perfect

Sharif000123 committed:
Formatted, one line was not perfect

Sharif000123 committed:
Fixed M_PI issue with #define _USE_MATH_DEFINES

Sharif000123 committed:
Fixed M_PI issue with #define _USE_MATH_DEFINES

Sharif000123 committed:
Commit

Sharif000123 committed:
Pushed some necessary changes (proposed by Max Ilsen) for the tests to run through

Sharif000123 committed:
Removing unnecessary includes

Sharif000123 committed:
Missed one include, added vector now

Sharif000123 committed:
Some very minor changes

Sharif000123 committed:
Added missing include in header file

Sharif000123 committed:
Added missing breakline at the end of the code

Sharif000123 committed:
Added some minor testing, as sonarcloud complained

Sharif000123 committed:
Some changes in the tests

Sharif000123 committed:
Fixed container issue with bounding-box pair, as it returns a pair of pairs, and fixed for scalar as well. Had to change some old code, for arithmetic types

Sharif000123 committed:
Modified Gabriel Ratio, because of one node graph crash

Sharif000123 committed:
corrected mistake in Layoutstatistics::angles (ata2(x,y) -> ata2(y,x)), made safety check in gabrielratio() by addin map from old to new graph, and added made centerof mass clearer

Sharif000123 committed:
Modified Layoutstatistics and layouthelper (moved graph init to above)

Sharif000123 committed:
Merge branch 'master' of https://github.com/ogdf/ogdf into layoutstatistics

Sharif000123 committed:
Rebased from master branch

Sharif000123 committed:
Added target_link_libraries to tests.cmake and commented out gabriel ratio test in Layoutstatistics, also for safety, I changed some functions used in gabriel ratio back to more basic options, as the other ones might have caused some issues

Sharif000123 committed:
Fixed Center of mass bug

Sharif000123 committed:
Gabriel Ratio now running

Sharif000123 committed:
Fixing formatter issue

Sharif000123 committed:
For safety, adapted all allDistances arrays to get passed in correct size

N-Coder commented on include/ogdf/basic/LayoutStatistics.h:
```diff
@@ -145,6 +213,231 @@ class OGDF_EXPORT LayoutStatistics {
+	 *
+	 * Returns an array of preservation values, of each node in Graph \p mainGraph
+	 */
+	static ArrayBuffer<double> neighbourhoodPreservation(const GraphAttributes& ga);
```
As this is a per-node statistic, use `NodeArray<double>` instead

N-Coder commented on src/ogdf/basic/LayoutStatistics.cpp:
```diff
@@ -40,13 +40,43 @@
+	return boundingBoxVals.width();
+}
+
+inline double graphArea(const GraphAttributes& ga) { return graphHeight(ga) * graphWidth(ga); }
```
Why do you compute the BoundingBox twice?

N-Coder commented on cmake/tests.cmake:
```diff
@@ -48,5 +48,6 @@ if(OGDF_SEPARATE_TESTS)
   endforeach()
 else()
   add_executable(tests EXCLUDE_FROM_ALL ${TEST_SOURCES} "${generated_path}")
+  target_link_libraries(tests OGDF COIN)
```
this should not be needed as it is handled by the line below

N-Coder reviewed (dismissed):
Some more comments, please also take into account the comments from my first review.

I debugged one of the failures for you and included a comment at the respective line.

In general, please prefer NodeArray and EdgeArray over ArrayBuffer as they automatically allocate the right size and check that you are using only the right nodes and edges of the Graph the array was registered for (so never use something like `nodearray[v->index()]` instead of `nodearray[v]`)

Sharif000123 committed:
Reverted Cmake file change

Sharif000123 committed:
Added proposed changes to Layoutstatistics, reworkined some of the methods, deleted bordercoordinates, reworked gabrielratio, and adapted layout_helpers to the changes

Sharif000123 committed:
Removed trailing whitespace

Sharif000123 committed:
Fixed bugs in layouthelpers, now it says error is in steiner-tree, but not sure if this is actually the problem

Sharif000123 commented:
Hi @N-Coder ,
I tried debugging by removing my new code step by step, until i nearly removed half the code, still getting the error, so i did the following: I've just deleted all my code, and pasted in the code from the master branch (into Layoutstatistics.cpp, Layoutstatistics.h, and layout_helpers.h to reverse all my changes) and deleted the build folder. I then fully built everything with CMAKE, and then run the test module via cmake, and I got the same error that I have gotten with my code. Can you please try yourself? Maybe switch to the master branch, and try what I did. Thanks

N-Coder commented:
Hi @Sharif000123, have you tried using a debugger? It usually tells you at which line the exception is thrown without having to change any code.

Sharif000123 commented:
> Hi @Sharif000123, have you tried using a debugger? It usually tells you at which line the exception is thrown without having to change any code.

Hi @N-Coder I tried debugging for several hours now, the debugger doesn't work as the tests file cant be created. I tried everything but I don't see a possibility to get the debugger to run on the tests.exe file as it cant be created. Did you try to build the tests file on the master branch?

Sharif000123 committed:
Reworked code, removed alldistances function as it was prone to errors

Sharif000123 committed:
Added missing catch in gabriel ration and better ones in edgeLengthDeviation

Sharif000123 committed:
removed usage of edgelengths function

Sharif000123 committed:
added some percautions in horizontalVerticalBalance

Sharif000123 committed:
Made neighbourhoodPreservation safe for multi-edges and self-loops

Sharif000123 committed:
aspect ratio had issues, built in an epsilon check for division of very small values

Sharif000123 committed:
changed all size_t to int to check for out of limit size_t subtraction

Sharif000123 committed:
fixed style mistake

Sharif000123 committed:
reversed size_t change and added one isfinite catch

Sharif000123 commented:
> This unfortunately sounds like a problem with your setup, as for me everything works fine on the master branch on my linux machine. Unfortunately, developing under Windows is always a _huge_ pain and I can't really help you with that (I never met a C++ expert that also knows Windows). Have you tried using WSL or MinGW?
> 
> I usually build with `OGDF_SEPARATE_TESTS` on so that I get separate executables for all the tests. For me it's just checkout the OGDF git in CLion and pass the cmake flags for debug mode (see below, but that's probably not the issue in your case with Windows being the problem) to set up, and then simply select the target I want to debug and hit the green 🐞 icon. 🤷🏼‍♂️
> 
> ```
> -DBUILD_SHARED_LIBS=ON
> -DOGDF_MEMORY_MANAGER=MALLOC_TS
> -DOGDF_LEAK_CHECK=ON
> -DOGDF_USE_ASSERT_EXCEPTIONS=ON
> -DOGDF_SEPARATE_TESTS=ON
> -DOGDF_USE_ASSERT_EXCEPTIONS_WITH_STACK_TRACE=ON_LIBUNWIND
> -DOGDF_WARNING_ERRORS=ON
> -DOGDF_DEBUG_MODE=HEAVY
> -DOGDF_INCLUDE_CGAL=ON
> -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
> ```
> 
> The steiner-tree tests seem to work with your latest commit (which is also true for the CI run), the trace I get for the failing planarization tests:
> 
> ```
> ogdf::internal::GraphList::head GraphList.h:324
> ogdf::internal::GraphList::begin GraphList.h:385
> ogdf::LayoutStatistics::gabrielRatio LayoutStatistics.cpp:515
> callLayout layout_helpers.h:280
> <lambda#1>::operator()(const ogdf::Graph &, const std::string &, const std::set<…> &) const layout_helpers.h:372
> std::__invoke_impl<…>(std::__invoke_other, <lambda#1> &, ogdf::Graph &, const std::string &, const std::set<…> &) invoke.h:63
> std::__invoke_r<…>(<lambda#1> &, ogdf::Graph &, const std::string &, const std::set<…> &) invoke.h:113
> std::_Function_handler::_M_invoke(const std::_Any_data &, ogdf::Graph &, const std::string &, const std::set<…> &) std_function.h:292
> std::function::operator() std_function.h:593
> <lambda#1>::operator()() const graphs.h:320
> ...
> ```
> 
> The issue is that the `allDistances` vector computed in `gabrielRatio` contains a single entry where both vertices are nullpointers. As I said, please replace this quadratic-sized array with ad-hoc computed distances. Similarly, please replace your ArrayBuffers with NodeArrays, as long as you simply use `ArrayBuffer::push` it will be impossible to look up the correct values for individual nodes.

Hi @N-Coder , 
can you try debugging the newest commit, I've fixed some issues and now most tests run, including all windows tests, but some linux and mac versions still spit an error, on a graph with 2 nodes and no edge, causing it to return nan, but i cant find it, as the prints are not working with this bug. Thanks in advance!

Sharif000123 committed:
Added some additional fixes and changed some affaybuffer functions to Edge-/NodeArray

Sharif000123 committed:
fixed issue with center of mass and closes pair of points functions

Sharif000123 committed:
fixed small mistakes

Sharif000123 committed:
fixed some casts

Sharif000123 committed:
Some concentration() fixes

Sharif000123 committed:
Stylistic changes

Sharif000123 committed:
Deleted unnecessary includes

Sharif000123 committed:
Added includes back

Sharif000123 committed:
Added missing includes for fix includes test, and removed unnecessary ones

Sharif000123 committed:
Undo last layoutstatistic.h include changes

Sharif000123 commented:
Hello @N-Coder ,
Right now its written that you requested changes, I don't see them though, the only thing I could do is request a review, if there is some change I have to do, please tell me. And thanks again for the assistance.

N-Coder commented:
GitHub refers to the inline comments I left in my two reviews above, most of which are by now outdated. Still, they need to be checked that the described issue is now fixed and then marked as *resolved*. 
Once that is done I can do a final pass, possibly also applying some of the fixes to the other legacy functions, and then merge this PR; although I can't promise that I'll get around to doing this this year.

Sharif000123 commented:
> GitHub refers to the inline comments I left in my two reviews above, most of which are by now outdated. Still, they need to be checked that the described issue is now fixed and then marked as _resolved_. Once that is done I can do a final pass, possibly also applying some of the fixes to the other legacy functions, and then merge this PR; although I can't promise that I'll get around to doing this this year.

@N-Coder  i've just marked all comments as resolved, but its still showing me that there are requested changes, do i need to reply to each comment so they are fully resolved?

N-Coder committed:
minor fixes

N-Coder committed:
Merge branch 'master' into layoutstatistics
